File: leetcode/253-meeting-rooms-ii/main.go
package main

import (
	"fmt"
	"sort"
)

type Interval struct {
	Start int
	End   int
}

/*
	1st approach
	- sort the array
	- create an array of meeting rooms
	- each interval, search for vacancies in the meeting rooms
	- if there is no collision, put the meeting in that room, else create a new meeting room for the interval
	Time		O(nlogn+n^2) nlogn: sort; n^2: search for vacancies by each interval
	Space 	O(n)	result array
	112ms, no beat because leetcode is recently updating the graph distribution
*/
func minMeetingRooms(intervals []Interval) int {
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i].Start < intervals[j].Start
	})
	rooms := [][]Interval{}
	for i := 0; i < len(intervals); i++ {
		j := 0
		for ; j < len(rooms); j++ {
			room := rooms[j]
			cop := []Interval{}
			cop = append(cop, room...)
			cop = append(cop, intervals[i])
			if canAttendMeetings(cop) {
				break
			}
		}
		if j < len(rooms) {
			rooms[j] = append(rooms[j], intervals[i])
		} else {
			rooms = append(rooms, []Interval{intervals[i]})
		}
	}
	return len(rooms)
}

func canAttendMeetings(intervals []Interval) bool {
	for i := 1; i < len(intervals); i++ {
		if intervals[i].Start < intervals[i-1].End {
			return false
		}
	}
	return true
}

func main() {
	a := []Interval{
		Interval{0, 30},
		Interval{5, 10},
		Interval{15, 20},
	}
	fmt.Println(minMeetingRooms(a))

	a = []Interval{
		Interval{7, 10},
		Interval{2, 4},
	}
	fmt.Println(minMeetingRooms(a))

	a = []Interval{
		Interval{0, 30},
		Interval{1, 30},
		Interval{2, 20},
		Interval{3, 30},
		Interval{4, 30},
		Interval{5, 20},
	}
	fmt.Println(minMeetingRooms(a))
}

File: leetcode/253-meeting-rooms-ii/main.py
import heapq

"""
    1st approach
    - sort the array
    - create an array to store the end time of the latest meeting in every room
    - each interval, compare the last meeting(since sorted) amongst the meeting in the meeting rooms
    - if there is no collision, put the meeting in that room, else create a new meeting room for the interval
    
    Time		O(NlogN + MN)   sort + iterate over the 'ends'
    Space 	    O(N)	        result array
    76 ms, faster than 52.42% 
"""


class Solution(object):
    def minMeetingRooms(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: int
        """
        if len(intervals) == 0:
            return 0

        def cptr(a, b):
            if a[0] == b[0]:
                return a[1]-b[1]
            return a[0]-b[0]
        intervals = sorted(intervals, cmp=cptr)

        ends = []
        for i in range(len(intervals)):
            s, e = intervals[i]
            found = False
            for j in range(len(ends)):
                if ends[j] <= s:
                    found = True
                    ends[j] = e
                    break
            if not found:
                ends.append(e)
        return len(ends)


"""
    2nd: sort + heap
    - sort the intervals by starting time
    - for each interval, compare the last meeting(since sorted) amongst the end times in the minheap
        - if there is no collision, put the meeting in that room
        - else create a new meeting room for the interval(by adding a new time in the minheap)

    Time    O(2NlogN)
    Space   O(N)
    52 ms, faster than 99.69%
"""


class Solution(object):
    def minMeetingRooms(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: int
        """
        intervals.sort()
        pq = []
        for s, e in intervals:
            if len(pq) > 0 and s >= pq[0]:
                pop = heapq.heappop(pq)
            heapq.heappush(pq, e)
        return len(pq)
